### PR TITLE
feat: support parse and parselazy with sax handler

### PR DESCRIPTION
**Pull request by liuq19**
# Main Changes:

1. refactor the parser as sax-style
2. parseondemand also use sax-style, and optimize lots

# ParseOnDemand Performance
```
Comparing /tmp/tmpzu6hwo3r.master.json to /tmp/tmpu_mulh5j.target.json
Benchmark                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------
twitter/SonicOnDemand_Normal                      -0.2318         -0.2318         26863         20637         26864         20636
citm_catalog/SonicOnDemand_Fronter                -0.8479         -0.8479         71691         10907         71690         10907
twitter/SonicOnDemand_NotFound                    -0.2406         -0.2406         26836         20379         26835         20379
tools2/draw-png.py /tmp/tmpu_mulh5j.target.json
```

# Parse Performance
```
Comparing /tmp/tmpxxfefh8b.master.json to /tmp/tmpmxor0ng0.target.json
Benchmark                                        Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------
book/Decode_SonicDyn                          +0.1570         +0.1570           688           796           688           796
twitter/Decode_SonicDyn                       -0.0024         -0.0024        137448        137114        137448        137114
poet/Decode_SonicDyn                          +0.0233         +0.0234        944678        966678        944547        966665
canada/Decode_SonicDyn                        +0.0054         +0.0054       2588317       2602310       2588272       2602258
github_events/Decode_SonicDyn                 +0.0310         +0.0311         31211         32178         31207         32178
lottie/Decode_SonicDyn                        -0.0081         -0.0081        680423        674913        680395        674906
otfcc/Decode_SonicDyn                         -0.0404         -0.0404     184449037     176991304     184448816     176991763
citm_catalog/Decode_SonicDyn                  -0.0160         -0.0160        880897        866765        880898        866769
twitterescaped/Decode_SonicDyn                +0.0187         +0.0187        380641        387773        380642        387768
fgo/Decode_SonicDyn                           -0.0583         -0.0585      67977428      64011544      67977512      63998854
gsoc-2018/Decode_SonicDyn                     +0.0103         +0.0104        801498        809738        801396        809740
```

### CODE REVIEW

**Comment by xiegx94** [comment]
Suqash all commits as one.

File: include/sonic/dom/handler.h [hunk 1]
```diff
@@ -15,26 +15,276 @@
...
+class SAXHandler {
+public:
+  using Allocator = typename NodeType::AllocatorType;
+
+  SAXHandler() = default;
+  SAXHandler(Allocator& alloc) : alloc_(&alloc) {}
+
+  SAXHandler(const SAXHandler&) = delete;
+  SAXHandler& operator=(const SAXHandler& rhs) = delete;
+  SAXHandler(SAXHandler&& rhs) : st_(rhs.st_), cap_(rhs.cap_), np_(rhs.np_), alloc_(rhs.alloc_) {
```
Comment (xiegx94) [inline include/sonic/dom/handler.h:40]: Format your code. This is larger than 80 chars

File: include/sonic/dom/parser.h [hunk 2]
```diff
@@ -34,280 +34,93 @@
...
+  sonic_force_inline ParseResult ParseLazy(const uint8_t* data, size_t len, LazySAX& sax) {
+    return parseLazyImpl(data, len, sax);
   }
 
  private:
-#define sonic_check_err()   \
-  if (err_ != kErrorNone) { \
-    return;                 \
+  sonic_force_inline bool hasTrailingChars() {
+    while (pos_ != len_) {
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:68]: This would be endless loop if pos_ is larger than len_.

File: include/sonic/dom/parser.h [hunk 3]
```diff
@@ -34,280 +34,93 @@
...
-      while (1) {
-        SONIC_REPEAT8(if (sonic_unlikely(*src == '\\')) break;
-                      else { *dst++ = *src++; });
-      }
-      goto cont;
+    uint8_t* sdst = src;
+    long n = internal::parseStringInplace(src);
+    pos_ = src - json_buf_;
+    if (n < 0) {
+      setParseError(static_cast<SonicError>(-n));
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:116]: Why parseStringInplace cannot return SonicError

File: include/sonic/dom/parser.h [hunk 4]
```diff
@@ -394,23 +208,25 @@ class Parser {
 
 #define SET_INT_AND_RETURN(int_val)                \
   {                                                \
-    JsonHandler<NodeType>::SetSint(node, int_val); \
+    if (!sax.Int(int_val)) RETURN_SET_ERROR_CODE(kParseErrorInvalidChar); \
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:211]: Format your codes

File: include/sonic/dom/parser.h [hunk 5]
```diff
@@ -643,220 +458,90 @@ class Parser {
...
-#define sonic_add_node()            \
-  {                                 \
-    if (sonic_likely(np_ < cap_)) { \
-      np_++;                        \
-    } else {                        \
-      goto err_invalid_char;        \
-    }                               \
+  template <unsigned parseFlags, typename SAX>
+  sonic_force_inline void parseImpl(SAX& sax) {
+#define sonic_check_err()   \
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:482]: Format

File: include/sonic/dom/parser.h [hunk 6]
```diff
@@ -643,220 +458,90 @@ class Parser {
...
-    size_t parent = 0;       // the postion of parent container node
+    // TODO (liuq19): vector is a temporary choice, will optimize in future.
+    std::vector<uint32_t> depth;
+    const uint32_t kArrMask = 1ull << 31;
+    const uint32_t kObjMask = 0;
 
     uint8_t c = scan.SkipSpace(json_buf_, pos_);
-    sonic_add_node();
-    switch (c) {
+   switch (c) {
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:494]: Format，indent

File: include/sonic/dom/parser.h [hunk 7]
```diff
@@ -643,220 +458,90 @@ class Parser {
...
-    parseString(st_[np_ - 1]);
+    parseStrInPlace(sax);
     sonic_check_err();
     c = scan.SkipSpace(json_buf_, pos_);
     if (sonic_unlikely(c != ':')) goto err_invalid_char;
 
     c = scan.SkipSpace(json_buf_, pos_);
-    sonic_add_node();
-    switch (c) {
+   switch (c) {
```
Comment (xiegx94) [inline include/sonic/dom/parser.h:528]: Format

File: include/sonic/dom/type.h [hunk 8]
```diff
@@ -22,7 +22,7 @@ namespace sonic_json {
 enum TypeFlag {
   // BasicType: 3 bits
   kNull = 0,    // xxxxx000
-  kNone = 1,    // xxxxx001
+  kRaw = 1,    // xxxxx001
```
Comment (xiegx94) [inline include/sonic/dom/type.h:25]: I suggest defining kRaw as 5

File: tests/merge_test.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      },
+  };
+
+  for (const auto& t : tests) {
+    auto ret = sonic_json::Merge(t.target, t.source);
+    EXPECT_STREQ(ret.c_str(), t.updated.c_str());
+  }
+}
+
+}
```
Comment (xiegx94) [inline tests/merge_test.cpp:120]: no newline at EOF

File: include/sonic/dom/type.h (hunk 8 above)
Comment (liuq19) [reply include/sonic/dom/type.h:25]: this is a basic type, are there more details?

**Review by xiegx94 — APPROVED** [review]
LGTM

**Pull request merged**